internal/client.go

```diff
@@ -626,6 +630,528 @@ type (
+	SchedulePolicies struct {
+		// Overlap - Controls what happens when an Action would be started by a Schedule at the same time that an older Action is still
+		// running.
+		Overlap enumspb.ScheduleOverlapPolicy
+
+		// CatchupWindow - The Temporal Server might be down or unavailable at the time when a Schedule should take an Action. When the Server
+		// comes back up, CatchupWindow controls which missed Actions should be taken at that point.
+		CatchupWindow time.Duration
+
+		// PauseOnFailure - When an Action times out or reaches the end of its Retry Policy.
+		PauseOnFailure bool
+	}
+
+	// ScheduleState describes the current state of a schedule.
+	ScheduleState struct {
+		// Note - Informative human-readable message with contextual notes, e.g. the reason
+		// a Schedule is paused. The system may overwrite this message on certain
+		// conditions, e.g. when pause-on-failure happens.
+		Note string
+
+		// Paused - True if the schedule is paused.
+		Paused bool
+
+		// LimitedActions - While true RemainingActions will be decremented for each action taken.
+		// Skipped actions (due to overlap policy) do not count against remaining actions.
+		LimitedActions bool
+
+		// RemainingActions - The Actions remaining in this Schedule. Once this number hits 0, no further Actions are taken.
+		// manual actions through backfill or ScheduleHandle.Trigger still run.
+		RemainingActions int64
+	}
+
+	// Schedule describes a created schedule.
+	Schedule struct {
+		// Action - Which Action to take
+		Action ScheduleAction
+
+		// Schedule - Describes when Actions should be taken.
+		Spec *ScheduleSpec
+
+		// SchedulePolicies - this schedules policies
+		Policy *SchedulePolicies
+
+		// State - this schedules state
+		State *ScheduleState
+	}
+
+	// ScheduleUpdate describes the desired new schedule from ScheduleHandle.Update.
+	ScheduleUpdate struct {
+		// Schedule - New schedule to replace the existing schedule with
+		Schedule *Schedule
+	}
+
+	// ScheduleUpdateOptions configure the parameters for updating a schedule.
+	ScheduleUpdateOptions struct{}
+
+	// ScheduleHandle represents a created schedule.
+	ScheduleHandle interface {
+		// GetID returns the schedule ID asssociated with this handle.
+		GetID() string
+
+		// Delete the Schedule 
+		Delete(ctx context.Context) error
+
+		// Backfill the schedule by going though the specified time periods and taking Actions as if that time passed by right now, all at once.
+		Backfill(ctx context.Context, backfill []ScheduleBackfill) error
+
+		// Update the Schedule. fn takes a description of the schedule and returns the new desired schedule.
+		// If update returns a nil response then no update will occur.
+		//
+		// NOTE: If two Update calls are made in parallel to the same Schedule there is the potential
+		// for a race condition.
+		Update(ctx context.Context, fn func(*ScheduleDescribeResponse) *ScheduleUpdate, options ScheduleUpdateOptions) error
```

Comment:
```suggestion
		Update(ctx context.Context, fn func(*ScheduleUpdate) error) error
```

```go
type ScheduleUpdate struct {
  // Read-only, alterations ignored
  Description *ScheduleDescribeResponse
  // Prepopulated from the description
  Schedule *Schedule
}
```

Benefits:

* Allows for more than just the describe response if we need it in the future
* Should prepopulate `Schedule` from the description. Args and memo and search attributes and what not can be `internal.EncodedValue` for pass-through needs.
* Removes need for "options" as the unwieldy last parameter after a closure. If/when a user wants to set options, we can expose those on the `ScheduleUpdate` such as fail-on-conflict vs retry-on-conflict
* Allows people to return an error from the callback

If you want people to be able to not update, I would support in order of preference: 1) allow `ErrSkipScheduleUpdate` to be returned, 2) changing return to `(bool, error)`, 3) have a `SkipUpdate bool` field on the update, or 4) attempt to diff against description to see if any mutations were made.

Reply:
>Should prepopulate Schedule from the description. Args and memo and search attributes and what not can be internal.EncodedValue for pass-through needs.

Probably misunderstanding, this doesn't seem like a benefit just a description of what the API is? The 1st and 4th benefit are clear I don't get why we need the prepopulated ` Schedule *Schedule` field since `Description .Schedule` would have the same information?

>Removes need for "options" as the unwieldy last parameter after a closure. If/when a user wants to set options, we can expose those on the ScheduleUpdate such as fail-on-conflict vs retry-on-conflict

I agree the last parameter options is unwieldy, I am not sure about exposing them in `ScheduleUpdate` since we may want to know an option before we run `fn` also has the possibility the options may change between calls. Maybe something like
```go
Update(ctx context.Context, options UpdateOptions) error

type UpdateOptions struct {
  
  Fn func(*ScheduleUpdate) error
  // example
  RetryCount int
}
```

Reply:
> I don't get why we need the prepopulated Schedule *Schedule field since Description .Schedule would have the same information?

If that remains the case once we solve the problem of two actions meaning the same, no problem. But we still might want to copy the pointer to top level so that they don't think the otherwise-read-only description is expected to be mutated.

> since we may want to know an option before we run fn

Usually I support making the API unwieldy for options support, this may be a bit much. I think your `RetryCount int` would be more like `RetryAttempt int` on `ScheduleUpdate` and the user could bail.

But I'm ok with what you have suggested there. I'd call the field `DoUpdate` instead of `Fn`, but if this single-required-field struct works for you, it works for me too.